Menus/UndoMenu/UndoMenu.cs

```diff
@@ -0,0 +1,204 @@
+            set
+            {
+                if (m_MenuHideFlags != value)
+                {
+                    m_MenuHideFlags = value;
+
+                    var visible = value == 0;
+                    if (m_UndoMenuUI)
+                        m_UndoMenuUI.visible = visible;
+
+                    if (visible)
+                        ShowFeedback();
+                    else
+                        this.ClearFeedbackRequests();
+                }
+            }
+        }
+
+        void Start()
+        {
+            m_UndoMenuUI = this.InstantiateUI(m_UndoMenuPrefab.gameObject).GetComponent<UndoMenuUI>();
+            m_UndoMenuUI.alternateMenuOrigin = alternateMenuOrigin;
+            m_UndoMenuUI.actions = menuActions;
+            this.ConnectInterfaces(m_UndoMenuUI); // Connect interfaces before performing setup on the UI
+            m_UndoMenuUI.Setup();
+            InputUtils.GetBindingDictionaryFromActionMap(m_ActionMap, m_Controls);
+            m_TrackpadController = this.GetDeviceType() == DeviceType.Vive;
+            m_FeedbackHintForCurrentController = m_TrackpadController
+                ? k_FeedbackHintForTrackpadController
+                : k_FeedbackHintForJoystickController;
+        }
+
+        public void ProcessInput(ActionMapInput input, ConsumeControlDelegate consumeControl)
+        {
+            var undoMenuInput = (UndoMenuInput)input;
+            if (undoMenuInput == null)
+            {
+                this.ClearFeedbackRequests();
+                return;
+            }
+
+            if (undoMenuInput.engage.wasJustReleased && !m_TrackpadController)
+                this.RestartCoroutine(ref m_StillEngagedAfterStickReleasedCoroutine, AcceptInputAfterStickReleased());
+
+            if (!(undoMenuInput.engage.wasJustPressed || !m_TrackpadController && (undoMenuInput.engage.isHeld || m_StillEngagedAfterStickRelease)))
+                return;
+
+            consumeControl(undoMenuInput.engage);
+            m_UndoMenuUI.engaged = true;
+
+            var navigateX = undoMenuInput.navigateX.value;
+            var undoRedoPerformed = false;
+            if (navigateX < -k_UndoRedoThreshold && (m_TrackpadController || m_PrevNavigateX > -k_UndoRedoThreshold))
+            {
+                Undo.PerformUndo();
+                m_UndoMenuUI.StartPerformedAnimation(true);
+                ShowUndoPerformedFeedback(true);
+                undoRedoPerformed = true;
+            }
+            else if (navigateX > k_UndoRedoThreshold && (m_TrackpadController || m_PrevNavigateX < k_UndoRedoThreshold))
+            {
+                Undo.PerformRedo();
+                m_UndoMenuUI.StartPerformedAnimation(false);
+                ShowUndoPerformedFeedback(false);
+                undoRedoPerformed = true;
+            }
+
+            m_PrevNavigateX = navigateX;
+
+            if (undoRedoPerformed)
+            {
+                consumeControl(undoMenuInput.navigateX);
```

Comment:
Should the AcceptInputAfterStickReleased coroutine be stopped here if currently running?